src/AzureExtension/Helpers/AzureRepositoryHierarchy.cs

```diff
@@ -0,0 +1,181 @@
+﻿// Copyright (c) Microsoft Corporation and Contributors
+// Licensed under the MIT license.
+
+using System.Collections.Concurrent;
+using DevHomeAzureExtension.Client;
+using DevHomeAzureExtension.DeveloperId;
+using Microsoft.TeamFoundation.Core.WebApi;
+using Microsoft.VisualStudio.Services.Account.Client;
+using Microsoft.VisualStudio.Services.Organization;
+using Microsoft.VisualStudio.Services.WebApi;
+
+// In the past, an organization was known as an account.  Typedef to organization to make the code easier to read.
+using Organization = Microsoft.VisualStudio.Services.Account.Account;
+
+namespace AzureExtension.Helpers;
+
+/// <summary>
+/// Handles the hierarchy between organizations and projects.  Handles querying for both as well.
+/// </summary>
+public class AzureRepositoryHierarchy
+{
+    private readonly ConcurrentDictionary<Organization, List<TeamProjectReference>> _organizationsAndProjects;
+
+    /// <summary>
+    /// Used to keep track of multiple Get requests to make sure work isn't duplicated.
+    /// </summary>
+    private readonly ConcurrentDictionary<Organization, Task<IPagedList<TeamProjectReference>>> _organizationsAndProjectTask;
+
+    private readonly DeveloperId _developerId;
+
+    /// <summary>
+    /// Used to keep track of all GetOrganization requests to make sure work isn't duplicated.
+    /// </summary>
+    private Task<List<Organization>>? _queryOrganizationsTask;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AzureRepositoryHierarchy"/> class.
+    /// Class to handle searching organizations and projects within a server.
+    /// </summary>
+    /// <param name="developerId">A developerId to an ADO instance.</param>
+    /// <remarks>
+    /// A server can have multiple organizations and each organization can have multiple projects.
+    /// Additionally, organizations and projects need to be fetched from the network.
+    /// This calss handles fetching the data, caching it, and searching it.
+    /// </remarks>
+    public AzureRepositoryHierarchy(DeveloperId developerId)
+    {
+        _developerId = developerId;
+        _organizationsAndProjects = new ConcurrentDictionary<Organization, List<TeamProjectReference>>();
+        _organizationsAndProjectTask = new ConcurrentDictionary<Organization, Task<IPagedList<TeamProjectReference>>>();
+    }
+
+    /// <summary>
+    /// Get all organizations the is apart of.
+    /// </summary>
+    /// <returns>A list of all organizations.</returns>
+    public async Task<List<Organization>> GetOrganizationsAsync()

```

Comment:
We can, if we wanted to stop the already running task.  Instead, the code will wait for the task to complete and use the results.  Why?

To get any information about repos.  Two calls need to happen.
1. Get all organizations.
2. Get all projects for the organizations.

If request 1 asks "Get me all repos."  The first this ADO will do is to get all organizations and projects.

While task 1 is getting all organizations, task 2 comes in and asks for repos, but with a search input.  Since task 1 is already doing the work to get the organizations and projects, task 2 can wait until task 1 is done.

If I canceled the work task 2 would need to query the server for the same information.